### ISSUE TITLE
NaNs when training

### ISSUE THREAD

**Issue by cooked-sashimi** [VCasecnikovs/Yet-Another-YOLOv4-Pytorch#16]
I have discovered that when training sometimes [line 1033](https://github.com/VCasecnikovs/Yet-Another-YOLOv4-Pytorch/blob/0b47389b81ec72823e64d5f4d097f6c9ce8626e2/model.py#L1033) 
```
pred_boxes[..., 2] = torch.exp(w) * self.anchor_w
```
or [line 1034](https://github.com/VCasecnikovs/Yet-Another-YOLOv4-Pytorch/blob/0b47389b81ec72823e64d5f4d097f6c9ce8626e2/model.py#L1034)
```
pred_boxes[..., 3] = torch.exp(h) * self.anchor_h
```
Causes overflow when there is an element in `w` or `h` that is too large. Is there any way to solve this?

**Comment by VCasecnikovs** [comment]
Hello, to be honest, I don't know how big should anchors be to cause overflowing. They should be enormously big...
Could You please provide an example of an image.

The model learns log of sizes, because it gives the model better control for enlargening.
That is why we should use exp for the outputs of the model.

**Comment by cooked-sashimi** [comment]
Hi @VCasecnikovs 

After debugging for a while I see that when I train:

`pred_boxes[..., 2].min()` and `pred_boxes[..., 3].min()` approaches 0

whereas: 

`pred_boxes[..., 2].max()` and `pred_boxes[..., 3].max()` approaches infinity

Though the loss continues to decrease, any ideas?

**Comment by VCasecnikovs** [comment]
Hello @cooked-sashimi
I think problem is with annotation.
Print targets.

**Comment by cooked-sashimi** [comment]
Here is this file that I am using:

[coco128.zip](https://github.com/VCasecnikovs/Yet-Another-YOLOv4-Pytorch/files/5662449/coco128.zip)

I convert the targets into yolo format before feeding them into the model, may this be the issue?

Here is an example of the target:

```
tensor([[0.0000e+00, 5.9000e+01, 5.1406e-01, 6.7969e-01, 6.6406e-01, 3.7500e-01],
        [0.0000e+00, 7.7000e+01, 5.5469e-01, 6.1563e-01, 1.1094e-01, 1.3594e-01],
        [0.0000e+00, 7.7000e+01, 4.5469e-01, 6.2656e-01, 9.6875e-02, 1.2656e-01],
        [0.0000e+00, 7.7000e+01, 5.1406e-01, 5.9844e-01, 1.9219e-01, 1.6875e-01],
        [0.0000e+00, 7.7000e+01, 6.1094e-01, 5.2812e-01, 1.1406e-01, 1.2500e-01],
        [0.0000e+00, 7.7000e+01, 6.8750e-01, 5.4062e-01, 5.3125e-02, 5.9375e-02],
        [0.0000e+00, 7.7000e+01, 3.6094e-01, 4.9531e-01, 1.6094e-01, 1.5000e-01],
        [0.0000e+00, 7.7000e+01, 5.4531e-01, 5.1562e-01, 6.7188e-02, 2.6563e-02],
        [0.0000e+00, 7.7000e+01, 6.5469e-01, 5.7344e-01, 9.0625e-02, 1.0312e-01],
        [0.0000e+00, 7.7000e+01, 7.3594e-01, 6.5000e-01, 2.1563e-01, 1.6094e-01],
        [0.0000e+00, 7.7000e+01, 5.6406e-01, 4.4688e-01, 7.6563e-02, 8.9062e-02],
        [1.0000e+00, 2.0000e+00, 5.8906e-01, 6.2500e-01, 6.5625e-02, 5.9375e-02],
        [1.0000e+00, 7.0000e+00, 9.1094e-01, 6.4844e-01, 1.2812e-01, 1.4531e-01]],
       device='cuda:0')
```

**Comment by cooked-sashimi** [comment]
Here is an example of the bboxes plotted on images
![image](https://user-images.githubusercontent.com/74332463/101702835-cd324180-3ad5-11eb-8b5c-db3ae0d5c943.png)

**Comment by VCasecnikovs** [comment]
Sorry, I have hard days. Will have a look at this tomorrow. Why you have non int classes?

**Comment by VCasecnikovs** [comment]
Also take a look at ylabels notation in the documentation.

**Comment by VCasecnikovs** [comment]
Have you tried to lower lr? Maybe there is a problem with gradients...

P.S. It writes your zip is dangerous.

**Comment by cooked-sashimi** [comment]
Yes I have tried to clip the gradients, it still happens.

I have realised that this comes down to the CIoULoss, as if I don't include the loss in the training then the predictions for the bounding boxes don't approach infinity.

Please try to download the data here: 
https://github.com/ultralytics/yolov5/releases/download/v1.0/coco128.zip

**Comment by VCasecnikovs** [comment]
What params are You using? It looks like you have problems with lr or optimizer or something else. I've trained on your dataset with default settings with the help of Training YOLOv4.ipynb and didn't have any problems. You should try to lower lr, maybe change optimizer. Do you use ListDataset to create dataset?

**Comment by VCasecnikovs** [comment]
You can print losses of CIOU loss. Shouldn't be so.

**Comment by cooked-sashimi** [comment]
Yes I have printed the CIOU loss, I have also tried to use MSE loss and it works fine, although the bbox predictions aren't very good. I do not use ListDataset to create dataset I will try that now

**Comment by cooked-sashimi** [comment]
@VCasecnikovs What hparams did you use, and how many classes did you use? I used 80, also what was your CIoU loss

**Comment by VCasecnikovs** [comment]
170 classes, more is ok (not needed would be 0 after some training), less is bad (you'll get some errors), but does it matter?) 
  hparams = {
      "n_classes" : 170,
  
      "pretrained" : False,
      "train_ds" : "train.txt",
      "valid_ds" : "valid.txt",
      "img_extensions" : [".JPG", ".jpg"],
      "bs" : 1,
      "momentum": 0.9,
      "wd": 0.001,
      "lr": 1e-8,
      "epochs" : 100,
      "pct_start" : 10/100,
      
      "optimizer" : "Ranger",
      "flat_epochs" : 50,
      "cosine_epochs" : 25,
      "scheduler" : "Cosine Delayed", 
      
      "SAT" : False,
      "epsilon" : 0.1,
      "SAM" : False,
      "ECA" : False,
      "WS" : False,
      "Dropblock" : False,
      "iou_aware" : False,
      "coord" : False,
      "hard_mish" : False,
      "asff" : False,
      "repulsion_loss" : False,
      "acff" : True,
      "bcn" : False,
      "mbn" : False,
  }

lr where change with lr find

**Comment by VCasecnikovs** [comment]
CIoU was starting from 90

**Comment by VCasecnikovs** [comment]
CIoUloss = (1 - iou_masked + rDIoU + alpha * v).sum(0)/num_samples

Num samples could be zero, so there must be a NaN.
I don't really know why I didn't have NaN here.

**Comment by VCasecnikovs** [comment]
The problem could be here,
`    CIoUloss = (1 - iou_masked + rDIoU + alpha * v).sum(0)/num_samples
`If num_samples is zero, answer would be inf
So I made a check.
Also, I made a check, what would be if we don't have y (lack of correct boxes).
No NaNs again.
```
  C: tensor([], grad_fn=<AddBackward0>); D: tensor([], grad_fn=<AddBackward0>)
  Confidence is object: 0.0, Confidence no object: 0.8261097073554993
  IoU: tensor([]); DIoU: tensor([], grad_fn=<DivBackward0>); alpha: tensor([]); v: tensor([], grad_fn=<MulBackward0>)
  CIoU : 0.0; Confindence: 82.61096954345703; Class loss should be because of label smoothing: 0.0
  C: tensor([], grad_fn=<AddBackward0>); D: tensor([], grad_fn=<AddBackward0>)
  Confidence is object: 0.0, Confidence no object: 0.6136282682418823
  IoU: tensor([]); DIoU: tensor([], grad_fn=<DivBackward0>); alpha: tensor([]); v: tensor([], grad_fn=<MulBackward0>)
  CIoU : 0.0; Confindence: 61.36282730102539; Class loss should be because of label smoothing: 0.0
  C: tensor([], grad_fn=<AddBackward0>); D: tensor([], grad_fn=<AddBackward0>)
  Confidence is object: 0.0, Confidence no object: 0.7705891132354736
  IoU: tensor([]); DIoU: tensor([], grad_fn=<DivBackward0>); alpha: tensor([]); v: tensor([], grad_fn=<MulBackward0>)
  CIoU : 0.0; Confindence: 77.05891418457031; Class loss should be because of label smoothing: 0.0
```

**Comment by cooked-sashimi** [comment]
Hi @VCasecnikovs I have tested with your script but after training for an amount of time the loss will still go to NaN. Please have a look [here](https://github.com/VCasecnikovs/Yet-Another-YOLOv4-Pytorch/pull/17)

**Comment by cooked-sashimi** [comment]
Do you use pretrained backbone?

**Comment by VCasecnikovs** [comment]
Yes, I do.

**Comment by VCasecnikovs** [comment]
Sorry, for testing Your dataset I didn't use pretrained backbone. It was written in config file above.

**Comment by VCasecnikovs** [comment]
I have looked at Your branch. It looks like exploding gradients.
There could be some problems, the first is: You use LR 6e-4, it could be too big for this task, so it causes gradients exploding. To prove this theory, there must be some NaN in outputs of the network.
Second, there could be some errors in CIoU loss, which make NaN, if it is so, please give me an output of the network and target with which NaN occurs.
Best regards,
Vadims.
